core/packet.h

```diff
@@ -0,0 +1,274 @@
+  return reinterpret_cast<Packet *>(rte_pktmbuf_alloc(ctx.pframe_pool()));
+}
+
+struct rte_mempool *get_pframe_pool();
+struct rte_mempool *get_pframe_pool_socket(int socket);
+
+void init_mempool(void);
+void close_mempool(void);
+
+// For the layout of snbuf, see snbuf_layout.h
+class Packet {
+ public:
+  struct rte_mbuf mbuf() {
+    return mbuf_;
+  }
+  void set_mbuf_data_off(uint16_t offset) { mbuf_.data_off = offset; }
+  void set_mbuf_data_len(uint16_t len) { mbuf_.data_len = len; }
+  void set_mbuf_pkt_len(uint32_t len) { mbuf_.pkt_len = len; }
+
+  Packet *vaddr() const { return vaddr_; }
+  void set_vaddr(Packet *addr) { vaddr_ = addr; }
+
+  phys_addr_t paddr() { return paddr_; }
+  void set_paddr(phys_addr_t addr) { paddr_ = addr; }
+
+  uint32_t sid() const { return sid_; }
+  void set_sid(uint32_t sid) { sid_ = sid; }
+
+  uint32_t index() const { return index_; }
+  void set_index(uint32_t index) { index_ = index; }
+
+  template <typename T = char *>
+  T reserve() {
+    return reinterpret_cast<T>(reserve_);
+  }
+
+  template <typename T = void *>
+  T head_data() {
+    return rte_pktmbuf_mtod(&mbuf_, T);
+  }
+
+  template <typename T = char *>
+  T data() {
+    return reinterpret_cast<T>(data_);
+  }
+
+  template <typename T = char *>
+  T metadata() {
+    return reinterpret_cast<T>(metadata_);
+  }
+
+  template <typename T = char *>
+  T scratchpad() {
+    return reinterpret_cast<T>(scratchpad_);
+  }
+
+  int head_len() { return rte_pktmbuf_data_len(&mbuf_); }
+
+  int total_len() { return rte_pktmbuf_pkt_len(&mbuf_); }
+
+  // single segment?
+  int is_linear() { return rte_pktmbuf_is_contiguous(&mbuf_); }
+
+  // single segment and direct?
+  int is_simple() { return is_linear() && RTE_MBUF_DIRECT(&mbuf_); }
+
+  void *prepend(uint16_t len) {
+    if (unlikely(mbuf_.data_off < len))
+      return nullptr;
+
+    mbuf_.data_off -= len;
+    mbuf_.data_len += len;
+    mbuf_.pkt_len += len;
+
+    return head_data();
+  }
+
+  // remove bytes from the beginning
+  void *adj(uint16_t len) {
+    if (unlikely(mbuf_.data_len < len))
+      return nullptr;
+
+    mbuf_.data_off += len;
+    mbuf_.data_len -= len;
+    mbuf_.pkt_len -= len;
+
+    return head_data();
+  }
+
+  // add bytes to the end
+  void *append(uint16_t len) { return rte_pktmbuf_append(&mbuf_, len); }
+
+  // remove bytes from the end
+  void trim(uint16_t to_remove) {
+    int ret;
+
+    ret = rte_pktmbuf_trim(&mbuf_, to_remove);
+    assert(ret == 0);
+  }
+
+  Packet *copy(Packet *src) {
+    Packet *dst;
+
+    assert(src->is_linear());
+
+    dst = __packet_alloc_pool(src->mbuf_.pool);
+
+    rte_memcpy(dst->append(src->total_len()), src->head_data(),
+               src->total_len());
+
+    return dst;
+  }
+
+  phys_addr_t dma_addr() { return seg_dma_addr(&mbuf_); }
+
+  // TODO: stream operator
+
+  void Dump(FILE *file);
+
+  static Packet *from_paddr(phys_addr_t paddr);
+
+  static inline int mt_offset_to_databuf_offset(
```

Comment:
If you are defining functions in the class declaration, you don't need to specify inline (because it will be, even if you don't want)